scripts/openwebnet-discovery.nse

```diff
@@ -0,0 +1,246 @@
+  ["MAC Address"] = "12",
+  ["Device Type"] = "15",
+  ["Firmware Version"] = "16",
+  ["Hardware Version"] = "17",
+  ["Uptime"] = "19",
+  ["Micro Version"] = "20",
+  ["Date and Time"] = "22",
+  ["Kernel Version"] = "23",
+  ["Distribution Version"] = "24",
+  ["Gateway IP address"] = "50",
+  ["DNS IP address 1"] = "51",
+  ["DNS IP address 2"] = "52"
+}
+
+local ACK = "*#*1##"
+local NACK = "*#*0##"
+
+-- Initiates a socket connection
+-- Returns the socket and error message
+local function get_socket(host, port, request)
+
+  local sd, response, early_resp = comm.opencon(host, port, request)
+
+  if sd == nil then
+    return nil, "Socket connection error."
+  end
+
+  if not response then
+    return nil, "Poor internet connection or no response."
+  end
+
+  if response == NACK then
+    return nil, "Received a negative ACK as response."
+  end
+
+  return sd, nil
+end
+
+local function get_response(sd, request)
+
+  local res = {}
+  local status, data
+
+  sd:send(request)
+
+  repeat
+    status, data = sd:receive_buf("##", true)
+    if status and data ~= ACK then
+      table.insert(res, data)
+    end
+    if data == ACK then
+      break
+    end
+
+    -- If response is NACK, it means the request method is not supported
+    if data == NACK then
+      res = {}
+    end
+  until not status
+
+  return res
+end
+
+local function format_dimensions(res)
+
+  if res["Date and Time"] then
+    res["Date"] = string.match(res["Date and Time"], "((%d+)%.(%d+)%.(%d+))$")
+
+    res["Time"] = string.match(res["Date and Time"], "^((%d+)%.(%d+)%.(%d+))")
+    res["Time"] = string.gsub(res["Time"], "%.", ":")
+
+    res["Date and Time"] = nil
+  end
+
+  if res["Device Type"] then
+    res["Device Type"] = device[ tonumber( res["Device Type"] ) ]
+  end
+
+  if res["MAC Address"] then
+    local t = {}
+    for _, v in ipairs(stdnse.strsplit("%.%s*", res["MAC Address"])) do
+      table.insert(t, stdnse.tohex(tonumber(v)))
+    end
+
+    res["MAC Address"] = table.concat(t, ":")
+  end
+
+  if res["Uptime"] then
+    local t = {}
+    local units = {
+      [0] = "d", "h", "m", "s"
+    }
+    local counter = 0
+
+    for _, v in ipairs(stdnse.strsplit("%.%s*", res["Uptime"])) do
+      table.insert(t, v .. units[counter])
+      counter = counter + 1
+    end
+
+    res["Uptime"] = table.concat(t, "")
+  end
+
+  return res
+
+end
+
+action = function(host, port)
+
+  local output = stdnse.output_table()
+
+  local sd, err = get_socket(host, port, ACK)
+
+  -- Socket connection creation failed
+  if sd == nil then
+    return err
+  end
+
+  -- Fetching list of dimensions of a device
+  for _, device in ipairs({"IP Address", "Net Mask", "MAC Address", "Device Type", "Firmware Version", "Uptime", "Date and Time", "Kernel Version", "Distribution Version"}) do
+
+    local head = "*#13**"
+    local tail = "##"
+
+    stdnse.debug("Fetching " .. device)
+
+    local res = get_response(sd, head .. device_dimension[device] .. tail)
```

Comment:
Need to check that `res` is valid here. A good test for a table with values in it is: `if res and next(res) then`